rfcs/test_harness/in_engine_test_harness.md

```diff
@@ -0,0 +1,402 @@
+- Larger security footprint
+
+#### Remote Console
+
+The Remote Console interface and CVars appear to already provide an interactive communication channel. Is RC deprecated? Should it be reused?
+
+### Removing files from deployable release builds
+
+- Is this feature necessary for a test framework?
+  - Should SIG-Content or SIG-Release own defining and maintaining this mechanism?
+- For ease of exclusion, should we recommend users compile their C++ tests into separate libraries, similar to existing C++ tests using GoogleTest?
+  - Would do little to simplify on platforms which require Monolithic builds.
+
+### Component Uniqueness
+
+- Can the "same" component exist between different levels, or are they functionally copies?
+  - Should a test ever be considered be the exact same test, if their environment is different?
+- What additional data is required for tracking test uniqueness in a debuggable, accessible way?
+  - What UX immediately gets the developer from a report to the test file?
+
+### Gem Name
+
+This feature needs an accessible, accurate, stable name for its Gem. The name should stay applicable even as other tools and test-features are enabled. The following names are proposed, but feedback is encouraged:
+
+- EngineTestHarness
+- TestHarness
+- GameTest
+- EngineTest
+
+### Multiple Gems
+
+Would splitting this feature across multiple gems to simplify how users select enabling features? Or would a single gem with automatic enable/disable behavior be better? (for example, what is the behavior when ScriptCanvas is disabled) The following layers could exist in separate gems:
+
+1. Standalone Remote Tool
+2. Test Harness Core (C++ interface and reporter, CTest helper, CLI runner)
+3. Lua Script Extension
+4. ScriptCanvas Extension
+5. Editor-based test report UI
+
+#### Reuse and Extension
+
+Aside from modifying the code of the Test Harness System, how could someone build other test tools from the individual parts of this harness? Inside O3DE this would likely be easier with multiple Gems communicating with AZ::Events. Outside of O3DE, it would be easier to have tools exist in their own GitHub repo, with a generic event/callback mechanism.
+
+### What UI avoids displaying too many test components?
+
+There should be few top-level menu options to add a generic test component to an entity, after which that component can then be configured to use any available test. The list of all tests, for which a project may contain hundreds, should not exist as top-level component options in the UI. Also, C++ tests should get configured similar to script-based tests, and not as individual unique components.
+
+### Could a test be written without opening the Editor?
+
+Tests attached to a component can be convenient, but require modifying entities in a level. It can also be convenient to provide a way to not require hooking up to an entity in a level. How could this provide a way to define and execute tests without manually opening the editor?
```

Comment:
Notably it is still possible to independently modify a script (or CPP file) after initially hooking it into a level in the editor.

This brings up interesting questions about what would be contained in a fully self-contained testing script:
* How would it indicate to the test system that it is an available test? via a manifest?
* Should it directly perform its own level loading? If not, how does it register external dependencies, such as a level it should be loaded into?
* Any entities it loads in would need to be in world-coordinates, which risks colliding with existing entities in a slightly different pattern than loading them relative to an existing "primary" component.